### ISSUE TITLE
Fix Waving Water Ambiguity and Adjust Default Height

### ISSUE THREAD

**Issue by benrob0329** [minetest/minetest#8369]
##### Issue type

 - Bug report
 - Feature request

 
##### Minetest version

```
5.0.0
```

##### OS / Hardware

Operating system: Arch Linux
CPU: Intel i7 5930k

GPU model: Nvidia 980ti
OpenGL version: 4.6.0

##### Summary

Waving water has long had issues with causing ambiguity with where the edge of water (or lava!). This bug has been around for years and is well known for causing people to not use waving water (myself included).

The main problem is that waving water recedes much further back than the edge of the water itself, while it should stay within the last flowing node.

![screenshot_20190312_174132](https://user-images.githubusercontent.com/19180682/54238453-e9631200-44ee-11e9-8d78-810b711142a4.png)

There is water all the way to the end of that channel, and most of the time the player has no idea unless they jump in (the same goes for lava!).

The second problem is much simpler, just a change in the default settings. When the waving water height is set to 1, you can see the underside of the wave surface, causing visual oddities due to semi-transparent textures.

![screenshot_20190312_174218](https://user-images.githubusercontent.com/19180682/54238711-950c6200-44ef-11e9-851f-ee72a06a71c7.png)

Simply changing this value to something like, `0.5` fixes the issue (although culling the underneath would be a proper fix).

![screenshot_20190312_174147](https://user-images.githubusercontent.com/19180682/54238600-4b237c00-44ef-11e9-9ffd-8e79117fa0c5.png)

**Comment by paramat** [comment]
I agree that the default waving water settings very probably need improving.

**Comment by paramat** [comment]
The current waving water settings seem ok to me, the surface variation stays within -0.5 to 0.0 nodes of non-waving surface.

I can't see a problem with those 'visual oddities', they don't look glitchy. However small the wave amplitude is you will always see wave undersides if you are low enough.
> The main problem is that waving water recedes much further back than the edge of the water itself, while it should stay within the last flowing node.

Unfixable, the last flowing node is very thin, so waving water inevitably makes it disappear.
The game could define a different water node so that placed water doesn't 'wave'.

**Comment by jastevenson303** [comment]
i use opaque water, so waving water is right out for me.  it shows thru the
world when it's opaque plus waving.

**Comment by paramat** [comment]
> it shows thru the world when it's opaque plus waving.

I tried but can't see a glitch, can you explain further?

**Comment by benrob0329** [comment]
>Unfixable, the last flowing node is very thin

Then don't make it so thin, or don't apply the "waviness" as much. I find it very hard to believe this "unfixable" claim when for some amount of time the waving animation is where it should be, and how this couldn't be modified to lessen the amount that the water recedes.

**Comment by jastevenson303** [comment]
>I tried but can't see a glitch, can you explain further?

My apologies, @paramat - either my recollection failed me, or it has since been modified and fixed - I seem to recall a change here, so I suspect the latter.

However, during my testing I was reminded that opaque waving leaves still exhibit a world see-through glitch.

**Comment by paramat** [comment]
I was wrong to say it's unfixable.
> Then don't make it so thin

The thinnest depth of flowing water cannot be made thick, it inevitably tapers to zero thickness, anything else would look ridiculous.

I also forgot that the waving water shader makes all liquids wave, it cannot be switched per-node as it needs to be, issue #5694 we need per-node `waving = 1` for liquids.

These issues seem to show inherent problems with waving water. Water should only wave like that in a large body of water like a lake or sea, but it waves in a single node hole too. Any liquid can be placed as a puddle of liquid somewhere and that waves too which looks weird. There's no way to decide if a liquid node is part of a large body of water.

For flowing water to not disappear when it is thin, we could try fixing it's lower face and modulate the height of the top face, but then that might make flowing water float above a source node and create a gap.
https://github.com/minetest/minetest/blob/e22a69d61a04624fe1eca818740b90dbb958e658/client/shaders/nodes_shader/opengl_vertex.glsl#L71

**Comment by pgimeno** [comment]
I've been toying with making the amplitude proportional to the thickness of the water layer, and got this patch:

```diff
diff --git a/client/shaders/nodes_shader/opengl_vertex.glsl b/client/shaders/nodes_shader/opengl_vertex.glsl
index 54b569c5e..c11922cea 100644
--- a/client/shaders/nodes_shader/opengl_vertex.glsl
+++ b/client/shaders/nodes_shader/opengl_vertex.glsl
@@ -70,9 +70,8 @@ float disp_z;
 
 #if (MATERIAL_TYPE == TILE_MATERIAL_LIQUID_TRANSPARENT || MATERIAL_TYPE == TILE_MATERIAL_LIQUID_OPAQUE) && ENABLE_WAVING_WATER
 	vec4 pos = gl_Vertex;
-	pos.y -= 2.0;
 	float posYbuf = (pos.z / WATER_WAVE_LENGTH + animationTimer * WATER_WAVE_SPEED * WATER_WAVE_LENGTH);
-	pos.y -= sin(posYbuf) * WATER_WAVE_HEIGHT + sin(posYbuf / 7.0) * WATER_WAVE_HEIGHT;
+	pos.y -= (2.0 + sin(posYbuf) * WATER_WAVE_HEIGHT + sin(posYbuf / 7.0) * WATER_WAVE_HEIGHT) * mod(pos.y - 0.002, 0.5) * 2.0;
 	gl_Position = mWorldViewProj * pos;
 #elif MATERIAL_TYPE == TILE_MATERIAL_WAVING_LEAVES && ENABLE_WAVING_LEAVES
 	vec4 pos = gl_Vertex;
```

Please give it a try. For what I've seen, it mostly solves the first issue. I didn't see any node where the node surface totally covered the water; every node showed at least a bit of water.

It works better to my taste with WATER_WAVE_LENGTH = 0.5, but not too bad with the default of 1.0.

**Comment by benrob0329** [comment]
That works much better, but there is still a bit where the water dips below the surface (it's only for a second though). It's a huge improvement, but I think it needs a bit more tweaking.

**Comment by jastevenson303** [comment]
@paramat Sorry I was wrong again, it does show through still.  Here's a video: https://www.youtube.com/watch?v=YDWrz8Bud2U

**Comment by pgimeno** [comment]
> That works much better, but there is still a bit where the water dips below the surface (it's only for a second though.

Yes, it goes slightly below the surface, but it's not too bad. This is the worst it got in my system. No node lacked water, which was my goal. With 0.5 amplitude it's even less noticeable.

![waving-water](https://user-images.githubusercontent.com/4267396/54277123-34ccfd00-458f-11e9-8fef-58787b02b49c.png)

As for the showing through, yes, it doesn't matter whether water is opaque or transparent: if a water source node has a solid node on top, then a gap is visible, because the water source node's top isn't rendered. No idea how to fix that.

![screenshot_20190313_130427](https://user-images.githubusercontent.com/4267396/54277782-123be380-4591-11e9-9f10-6906bea5669d.png)

**Comment by benrob0329** [comment]
That would require the water surface to always be drawn, even with a node directly on top.

@pgimeno the biggest problem I can see is that now there is water disconnected from the main stream, which looks wrong.

**Comment by paramat** [comment]
With pgimeno's patch:

![screenshot_20190313_214127](https://user-images.githubusercontent.com/3686677/54316842-9ade7200-45d9-11e9-846a-478153285f88.png)

^ When thinnest flowing water is above source nodes, big step-downs, not a big issue.

![screenshot_20190313_214346](https://user-images.githubusercontent.com/3686677/54316919-bfd2e500-45d9-11e9-85fb-94b6ec85c894.png)

^ Lots of z-fighting flicker on the pointed node face.

**Comment by pgimeno** [comment]
OK, I've figured it out. The correct modulo was 10.0 and there was an offset of 5.0. Which leads to this patch:

```diff
diff --git a/client/shaders/nodes_shader/opengl_vertex.glsl b/client/shaders/nodes_shader/opengl_vertex.glsl
index 54b569c5e..df7b8e722 100644
--- a/client/shaders/nodes_shader/opengl_vertex.glsl
+++ b/client/shaders/nodes_shader/opengl_vertex.glsl
@@ -70,9 +70,8 @@ float disp_z;
 
 #if (MATERIAL_TYPE == TILE_MATERIAL_LIQUID_TRANSPARENT || MATERIAL_TYPE == TILE_MATERIAL_LIQUID_OPAQUE) && ENABLE_WAVING_WATER
 	vec4 pos = gl_Vertex;
-	pos.y -= 2.0;
 	float posYbuf = (pos.z / WATER_WAVE_LENGTH + animationTimer * WATER_WAVE_SPEED * WATER_WAVE_LENGTH);
-	pos.y -= sin(posYbuf) * WATER_WAVE_HEIGHT + sin(posYbuf / 7.0) * WATER_WAVE_HEIGHT;
+	pos.y -= (2.0 + (sin(posYbuf) + sin(posYbuf / 7.0)) * WATER_WAVE_HEIGHT) * (mod(pos.y + 4.998, 10.0) * 0.1);
 	gl_Position = mWorldViewProj * pos;
 #elif MATERIAL_TYPE == TILE_MATERIAL_WAVING_LEAVES && ENABLE_WAVING_LEAVES
 	vec4 pos = gl_Vertex;
```

This one decreases the wave height according to the original node height. When it's too low, the wave height is almost 0.

**Comment by paramat** [comment]
Yes that's better, the issues i show remain but personally don't bother me, waving water looks bad as it currently is anyway.

**Comment by benrob0329** [comment]
New patch looks good.

I will note that Modulo 11 is more what I had in mind for keeping the general shore "waviness" of the original code while always keeping water at the tip of the shore, but I'd be happy to see a merged PR with either.
![screenshot_20190314_000841](https://user-images.githubusercontent.com/19180682/54330826-9b810380-45ed-11e9-9cd6-1a06c93726f2.png)

**Comment by pgimeno** [comment]
My understanding is that modulo 11 would make it look different depending on the Y position of the water, which I don't consider acceptable. Note that the node coordinates are multiplied by 10 when they get to the vertex shader, therefore calculating the modulo 10 gives the Y position within the node from 0 to 9.999999, which is then divided by 10 (actually multiplied by 0.1) to give the factor for the wave amplitude. Values like the 0.5 I used earlier worked because 10 is a multiple of 0.5; but a modulo of 11 would not be in phase with the vertical position, making it look different at other heights and breaking the effect.

If you're after having a minimum waviness, it could be accomplished by adding a constant to the modulo. Something like this (untested) should keep a ~10% minimum waviness:

```glsl
float min_waviness = 1.111111;
float amplitude_factor = (mod(pos.y + 4.998, 10.0) + min_waviness) / (10.0 + min_waviness)
pos.y -= (2.0 + (sin(posYbuf) + sin(posYbuf / 7.0)) * WATER_WAVE_HEIGHT) * amplitude_factor;
```

**Comment by benrob0329** [comment]
+1 to a `min_waviness` setting.

In the long run, this shader should look at the nodedef for a `waving=N` (where N is multiplied by `WATER_WAVE_HEIGHT` or something)

**Comment by lhofhansl** [comment]
There was some discussion on IRC about this.
The discussion here was dormant for 9 months. Let's warm it back up again. We have since replaced the sin waves with Perlin-type noise waves. And for lava we have since "fixed" it, by just having no waves lava (lava is not wavy).

It seems what we want to taper waves off based on the "thinness" of the node.
Another approach is make the height conditional to how deep the water is (along the y-axis).
Yet another is to take the surrounding nodes into account too (along the x and z axis').

Neither of those would look very realistic. In nature waves are typically longitudinal and fast in deep water and slow down and turn "up" (transversal) only in shallow water. Of course that's not generally true either, there are huge waves in the middle of the ocean.

The modulo approach here seems fragile, relying too much of "numerical quirks" :)

In the end a wave is a moving height difference of the water surface, that height difference has to be expressed visually somewhere. If that's not wanted, just turn waving water off...?

I'll also try another approach, if we have flowing water, would could multiply the configured wave height with with the thickness of the node, and hence the waves would turn more shallow there - the trick would probably be to make the waves smooth still.

**Comment by paramat** [comment]
Since any damaging liquid can be fixed by not making it wave, this issue is surely low priority now?
So far the solutions have just made liquids look worse and added complexity, so i am not convinced it is worth it.

**Comment by benrob0329** [comment]
This is probably low-priority, but I would still like to see the min_waviness value be a part of the shader.
